### PR TITLE
Monitor pod status in submission v2.

### PR DESCRIPTION
**Pull request by mccheah**
Allows the submission client to remain running until the driver pod completes. This comprises a number of refactors to allow the pod monitoring to be mocked in a unit test, and also includes a variant of #282 so that both V1 and V2 can have the exit code of the containers be logged at the end of the running application.

### CODE REVIEW

**Comment by mccheah** [comment]
@lins05  @ash211

**Comment by mccheah** [comment]
Note that this approach differs in some key ways from #282. The statuses of all containers are logged without necessarily assuming the driver container is the only container, although it's more than likely that there will only be one container here. The status of the containers are described without an assumption that they are necessarily in the "FINISHED" state. Finally, the logging itself is built into the new `LoggingPodStatusWatcher#awaitCompletion` API which also makes it such that callers do not need to maintain their own `CountDownLatch` as well as not needing to worry about separately asking the Watcher to log the final status.

It would be good to compare this with #282 and combine the best aspects of each.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala
Comment (lins05) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:139]: +1 for encapsulating the latch here

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (lins05) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: Is this necessary? I see yarn also use this https://github.com/apache/spark/blob/v2.1.0/yarn/src/main/scala/org/apache/spark/deploy/yarn/config.scala#L134

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala
Comment (lins05) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:21]: This is exceeds 100 chars and would fail scala style check.

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:21]: Scalastyle doesn't check imports for line length.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: Good point - reverted it.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:175]: where would this exception get caught?  I don't think we want to throw an exception that possibly blocks application flow just for logging purposes -- maybe instead log Container state "Unknown" so as to not break code flow

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v1/Client.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v1/Client.scala]: seems like this check should be done inside the LoggingPodStatusWatcherImpl -- otherwise all clients of that API need to do this check

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/Client.scala
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/Client.scala]: should we put similar logging for application finished / launched / waiting on completion here as we have in V1?  I rather like the progress logging for debugging purposes (and building an intuition for where time is spent in the process)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:175]: This should be happening in the logging thread, I don't think this will crash the application.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/Client.scala
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/Client.scala]: One difference here is that we create all of the resources with fire-and-forget semantics if SparkSubmit doesn't wait for the application to complete. The pod is created and SparkSubmit exits without checking if the pod has started running. This is because unlike the first version of submission which requires the driver to be running in order to bootstrap it, the bootstrap process here can be done without any additional interaction from the submission client other than to create the resources. However it's not clear if the submission client should also wait for the driver pod to start running before exiting even if fire-and-forget mode is turned on.

Comment (ash211) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/v2/Client.scala]: I _think_ the semantics of fire-and-forget is that you have to make sure it's actually fired before forgetting.  Do you know if in yarn-cluster mode the submission fails if e.g. the job gets hung in ACCEPTED state and the AM never gets resources?

This is probably something we can do in a followup commit and needn't block this one

**Review by ash211 — APPROVED** [review]
LGTM, will merge when build passes

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala
Comment (kimoonkim) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:140]: Shouldn't this be `podCompletedFuture.await`? My client exits right away without waiting for the driver pod to complete. I wonder if it's because of this line?

>2017-06-02 14:48:36 INFO  Client:54 - Waiting for application spark-hdfstest-1496440110123 to finish...
2017-06-02 14:48:36 INFO  LoggingPodStatusWatcherImpl:54 - Container final statuses:
	 Container name: spark-kubernetes-driver
	 Container image: docker:5000/spark-driver:kimoon-0602-1
	 Container state: Waiting
	 Pending reason: PodInitializing
2017-06-02 14:48:36 INFO  Client:54 - Application spark-hdfstest-1496440110123 finished.

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:140]: Good catch, missed this. Can you send a fix?

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/LoggingPodStatusWatcher.scala:140]: Never mind I see the fix now.

**Pull request merged**
